finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/BufferedStream.scala

```diff
@@ -0,0 +1,212 @@
+
+  def bufferSize: Long = _bufferSize
+
+  def onEnd: Future[Unit] = if (underlying.isEmpty)
+    Future.Unit
+  else
+    _onEnd
+
+  /**
+   * Attempt to create a child Stream.  If the buffer has not yet been discarded, returns a
+   * Stream and offers the Frames in the buffer to that Stream.  All further Frames that are read
+   * from the underlying Stream will offered to that Stream as well.  If the buffer has been
+   * discarded, it is impossible to create a child Stream since the beginning of the Stream is
+   * lost.  In this case, return a Throw.
+   */
+  def fork(): Try[Stream] = {
+    state match {
+      case State.Buffering => synchronized {
+        val q = new AsyncQueue[Frame]()
+
+        for (f <- buffer) {
+          f.open()
+          // Any instance of RefCountedFrame must also be an instance of Frame so this cast should
+          // be safe.
+          q.offer(f.asInstanceOf[Frame])
+        }
+        forks += q
+        val stream = if (underlying.isEmpty) {
+          Stream.empty(q)
+        } else {
+          new AsyncQueueReader { child =>
+            override protected[this] val frameQ = q
+            override def read(): Future[Frame] = child.synchronized {
+              // If the queue is empty, pull a Frame from the underlying Stream (which will be
+              // offered to all child Streams) before polling the queue.
+              if (q.size == 0) bufferedStream.pull()
+              super.read()
+            }
+          }
+        }
+        Return(stream)
+      }
+      case e: Exception => Throw(e)
+    }
+  }
+
+  private[this] val bufferDiscardedP = new Promise[Unit]
+  val onBufferDiscarded: Future[Unit] = bufferDiscardedP
+
+  /**
+   * Clear the buffer and release all the Frames.  Note that the Frames are reference counted so
+   * each underlying Frame will not be released until all child Streams have released the Frame as
+   * well.  After this is called, it will no longer be possible to create child Streams.
+   */
+  def discardBuffer(): Unit = synchronized {
+    if (state == State.Buffering) {
+      state = State.BufferDiscarded
+      for (bufferFrame <- buffer) bufferFrame.release()
+      buffer.clear()
+      val _ = bufferDiscardedP.setDone()
+    }
+  }
+
+  /**
+   * Read a Frame from the underlying Stream and write it into the buffer and offer it to the child
+   * Streams.
+   */
+  private[this] def pull(): Future[Unit] = {
+    if (underlying.isEmpty) {
+      Future.Unit
+    } else {
+      underlying.read().transform {
+        case Return(f: Frame.Data) => synchronized {
+          val refCounted = new RefCountedDataFrame(f)
+          handleFrame(refCounted, f.buf.length)
+          if (f.isEnd) _onEnd.setDone()
+          Future.Unit
+        }
+        case Return(f: Frame.Trailers) => synchronized {
+          val refCounted = new RefCountedTrailersFrame(f)
+          // TODO: Determine size of Trailers Frame to count against buffer capacity
```

Comment:
fwiw, for the purposes of HTTP/2, only data frames are counted against flow control.  so this probably doesn't have to be todone